src/myrtlespeech/run/run.py

```diff
@@ -47,72 +49,255 @@ def __call__(self, sentence: List[str]) -> List[str]:
         return new_sentence
 
 
-class ReportCTCDecoder(Callback):
-    """TODO
+class ReportDecoderBase(Callback):
+    """For reporting error rates word and character error rates.
+
+    .. note::
+
+        This class should not be instantiated directly.
+
+    When overriding the base class, you should define a
+    :py:meth:`get_decoder_inputs` method so that the decoder is able to
+    access the required inputs.
+
 
     Args:
-        ctc_decoder: decodes output to sequence of indices based on CTC
+        decoder: decodes output to sequence of indices.
+
+        alphabet: converts sequences of indices to sequences of symbols
+            (:py:data:`strs`).
+
+        word_segmentor: groups sequences of symbols into sequences of words.
+            By default this splits words on the space symbol.
+
+        eval_every: WER/CER is cacluated every ``eval_every``th epoch. Default
+            is 1.
+
+        calc_quantities: Iterable of strings of error quantities to calculate.
+            The strings can take the following values:
 
-        alphabet: converts sequences of indices to sequences of symbols (strs)
+                'wer': Word-error rate is calculated.
 
-        word_segmentor: groups sequences of symbols into sequences of words
+                'cer': Character-error rate is calculated.
+
+            ``calc_quantities`` defaults to ('cer', 'wer').
     """
 
-    def __init__(self, ctc_decoder, alphabet, word_segmentor):
-        self.ctc_decoder = ctc_decoder
+    def __init__(
+        self,
+        decoder,
+        alphabet,
+        word_segmentor=WordSegmentor(" "),
+        eval_every=1,
+        calc_quantities=("cer", "wer"),
+    ):
+        self.decoder = decoder
         self.alphabet = alphabet
         self.word_segmentor = word_segmentor
+        self.eval_every = eval_every
+
+        for idx, error_rate in enumerate(calc_quantities):
+            assert error_rate in [
+                "cer",
+                "wer",
+            ], f"calc_quantities[{idx}]={error_rate} is not in ['cer', 'wer']."
+        assert (
+            list(set(calc_quantities)).sort() == list(calc_quantities).sort()
+        ), f"Repeated error rate in {calc_quantities}"
+        assert len(calc_quantities) > 0, "calc_quantities cannot be empty!."
+        self.calc_quantities = list(calc_quantities)
```

Comment:
This is a good place to use an enum: 

https://docs.python.org/3/library/enum.html

e.g.

```python3

class Metric(Enum):
    WER = 0
    CER = 1

def __init__(...
    metrics: Tuple[Metric] = (Metric.WER, Metric.CER)
):
```

Reply:
Good point - I've used one. Thx